src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -90,70 +106,68 @@ struct TensorContract
                 "Cannot contract the requested indices.");
 
-  using new_type = detail::ComputeContractedType<Index1, Index2, T, X, Symm,
-                                                 IndexList, ArgsList>;
+  using new_type = detail::ComputeContractedType<ReplacedArg1, ReplacedArg2, T,
+                                                 X, Symm, IndexList, ArgsList>;
 
   using type = X;
   using symmetry = typename new_type::symmetry;
   using index_list = typename new_type::index_list;
-  static constexpr auto num_tensor_indices =
-      tmpl::size<index_list>::value == 0 ? 1 : tmpl::size<index_list>::value;
-  using args_list = tmpl::sort<typename new_type::args_list>;
+  static constexpr auto num_tensor_indices = tmpl::size<index_list>::value;
+  using args_list = typename new_type::args_list;
 
   explicit TensorContract(
       const TensorExpression<T, X, Symm, IndexList, ArgsList>& t)
       : t_(~t) {}
 
-  template <size_t I, size_t Rank, Requires<(I <= Index1::value)> = nullptr>
-  SPECTRE_ALWAYS_INLINE void fill_contracting_tensor_index(
-      std::array<int, Rank>& tensor_index_in,
-      const std::array<int, num_tensor_indices>& tensor_index) const {
-    // -100 is for the slot that will be set later. Easy to debug.
-    tensor_index_in[I] = I == Index1::value ? -100 : tensor_index[I];
-    fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
-  }
-
-  template <size_t I, size_t Rank,
-            Requires<(I > Index1::value and I <= Index2::value and
-                      I < Rank - 1)> = nullptr>
-  SPECTRE_ALWAYS_INLINE void fill_contracting_tensor_index(
-      std::array<int, Rank>& tensor_index_in,
-      const std::array<int, Rank - 2>& tensor_index) const {
-    // tensor_index is Rank - 2 since it shouldn't be called for Rank 2 case
-    // -200 is for the slot that will be set later. Easy to debug.
-    tensor_index_in[I] = I == Index2::value ? -200 : tensor_index[I - 1];
-    fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
-  }
-
-  template <size_t I, size_t Rank,
-            Requires<(I > Index2::value and I < Rank - 1)> = nullptr>
+  template <size_t I, size_t Rank>
   SPECTRE_ALWAYS_INLINE void fill_contracting_tensor_index(
-      std::array<int, Rank>& tensor_index_in,
-      const std::array<int, Rank - 2>& tensor_index) const {
-    // Left as Rank - 2 since it should never be called for the Rank 2 case
-    tensor_index_in[I] = tensor_index[I - 2];
-    fill_contracting_tensor_index<I + 1>(tensor_index_in, tensor_index);
-  }
-
-  template <size_t I, size_t Rank, Requires<(I == Rank - 1)> = nullptr>
-  SPECTRE_ALWAYS_INLINE void fill_contracting_tensor_index(
-      std::array<int, Rank>& tensor_index_in,
-      const std::array<int, num_tensor_indices>& tensor_index) const {
-    tensor_index_in[I] = tensor_index[I - 2];
+      std::array<size_t, Rank>& tensor_index_in,
+      const std::array<size_t, num_tensor_indices>& tensor_index) const {
```

Comment:
It's worth mentioning here that the type stored by the arrays here is `size_t`, but the 2nd parameter passed in in the call to `fill_contracting_tensor_index` that is done in the multi-index `get` function has a templated type `U`, not specifically `size_t`. The reason that the types of these two arrays were changed from `int` to `size_t` was because:
- The `tensor_index` array in the multi-index `get` should store the same type as the argument array (`new_tensor_index`) of the multi-index `get`. This is because in `fill_contracting_tensor_index`, the former's elements are assigned using values from the latter.
- This `tensor_index` is then passed to a call to  `ComputeContractionImpl::apply` in the multi-index `get` function, where it's passed to another call to `get`
- This calls `TensorExpression:get`, which calls `Tensor::get`, which requires that the tensor multi-index provided (e.g. a `std::array` like `[1, 2, 3]` for some component R{123} of some tensor R), the type of the array must be `size_t` in order to call `Structure::get_storage_index`.
- Therefore, the types of `new_tensor_index` and `tensor_index` in the multi-index `get` must be the same (so we can assign the former's values to the latter), but also, `tensor_index` must store `size_t` because of this call to `Tensor::get` calling `Structure` functions that require `tensor_index` stores `size_t`

As it is now, it works assuming `size_t` because that's all that's currently used for these multi-index values. In other words, that `U` template param in the multi-index `get` being `size_t` is the only instantiation right now (I believe). However, this of course isn't a great idea to just assume/rely on `U == size_t`. Some thoughts:
- Maybe the `U` template parameter can be moved from the multi-index `get` altogether, and `size_t` can be used as the type of the parameter (`new_tensor_index`). If the values in the multi-index will always be positive, I think this would be a good move. This also means that the `U` template param should be removed for `AddSub::get`, as well, and `size_t` likewise used in its place. It's important to re-note that functions in `Structure.hpp` require multi-indices to have `size_t` values
- If not, then `tensor_index` needs to store `size_t` but still be able to safely be assigned values from `new_tensor_index`, even though we can't count on its type also being `size_t`

Reply:
@nilsdeppe Let me know if you have thoughts on this. And if it's easier to discuss over slack / video chat, that works, too